pkg/syncer/namespace/namespace_downstream_controller.go

```diff
@@ -100,29 +99,45 @@ func NewDownstreamController(
 			return downstreamClient.Resource(namespaceGVR).Delete(ctx, namespace, metav1.DeleteOptions{})
 		},
 		upstreamNamespaceExists: func(clusterName logicalcluster.Name, upstreamNamespaceName string) (bool, error) {
-			_, err := upstreamInformers.ForResource(namespaceGVR).Lister().ByCluster(clusterName).Get(upstreamNamespaceName)
+			lister, known, synced := ddsifForUpstreamSyncer.Lister(namespaceGVR)
+			if !known || !synced {
+				return false, errors.New("informer should be up and synced for namespaces in the upstream syncer informer factory")
+			}
+			_, err := lister.ByCluster(clusterName).Get(upstreamNamespaceName)
 			if apierrors.IsNotFound(err) {
 				return false, nil
 			}
-			return !apierrors.IsNotFound(err), err
+			if err != nil {
+				return false, err
+			}
+			return true, nil
 		},
 		getDownstreamNamespace: func(downstreamNamespaceName string) (runtime.Object, error) {
-			return downstreamInformers.ForResource(namespaceGVR).Lister().Get(downstreamNamespaceName)
+			lister, known, synced := ddsifForDownstream.Lister(namespaceGVR)
+			if !known || !synced {
+				return nil, errors.New("informer should be up and synced for namespaces in the downstream informer factory")
+			}
+			return lister.Get(downstreamNamespaceName)
 		},
 		listDownstreamNamespaces: func() ([]runtime.Object, error) {
-			return downstreamInformers.ForResource(namespaceGVR).Lister().List(labels.Everything())
+			lister, known, synced := ddsifForDownstream.Lister(namespaceGVR)
+			if !known || !synced {
+				return nil, errors.New("informer should be up and synced for namespaces in the downstream informer factory")
+			}
+			return lister.List(labels.Everything())
 		},
 		isDowntreamNamespaceEmpty: func(ctx context.Context, namespace string) (bool, error) {
```

Comment:
unrelated to this PR: this function sounds dangerous. What is it used for?

Reply:
It is used as a condition to allow the deletion of a downstream namespace. Downstream namespace deletion should only occur (after a delay) when no more synced resources exist inside it.